# SQL CARTO QUERY FUNCTION
# written for PUI2016_Python2

##### This notebook is set up to link as a default to the fb55 account.  To turn in the homework use the same account you used in the lab (hvt201)  and query the database that you were querying in class (citibike_feb_2015)

In [2]:
SQL_SOURCE = 'https://ekh331.carto.com/api/v2/sql?q='

import urllib2
import urllib
import StringIO
import ast
import pandas as pd

def queryCartoDB(query, format='CSV', source=SQL_SOURCE):
    '''queries carto datasets from a given carto account
    Arguments: 
    query - string: a valid sql query string
    format - outlut format  OPTIONAL (default CSV)
    source - a valid sql api endpoint OPTIONAL (default carto fb55 account)
    Returns:
    the return of the sql query AS A STRING
    NOTES:
    designed for the carto API, tested only with CSV return format'''
    
    data = urllib.urlencode({'format': format, 'q': query})
    try:
        response = urllib2.urlopen(source, data)
    except urllib2.HTTPError, e:
        raise ValueError('\n'.join(ast.literal_eval(e.readline())['error']))
    except Exception:
        raise
    return response.read()

In [3]:
test = '''
SELECT *
FROM citibike
'''

In [4]:
pd.read_csv(StringIO.StringIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0            5960       NaN                   NaN       1      1965.0   19217   
1            9806       NaN                   NaN       1      1971.0   17575   
2           13113       NaN                   NaN       1      1973.0   20313   
3               1       NaN                   NaN       2      1978.0   17131   
4               2       NaN                   NaN       1      1993.0   21289   
5               3       NaN                   NaN       2      1969.0   18903   
6               4       NaN                   NaN       2      1985.0   21044   
7               5       NaN                   NaN       1      1957.0   19868   
8               6       NaN                   NaN       1      1979.0   15854   
9               7       NaN                   NaN       1      1983.0   15173   
10              8       NaN                   NaN       1      1955.0   17862   
11              9       NaN                   NaN       2      1985.0   21183   
12             10       NaN                   NaN       1      1982.0   14843   
13             11       NaN                   NaN       2      1988.0   16936   
14             12       NaN                   NaN       2      1991.0   21507   
15             13       NaN                   NaN       2      1982.0   17456   
16             14       NaN                   NaN       2      1990.0   14679   
17             15       NaN                   NaN       1      1977.0   18978   
18             16       NaN                   NaN       1      1968.0   16496   
19             17       NaN                   NaN       1      1990.0   17107   
20             18       NaN                   NaN       1      1987.0   16984   
21             19       NaN                   NaN       1      1980.0   15997   
22             20       NaN                   NaN       1      1960.0   21516   
23             21       NaN                   NaN       1      1986.0   15815   
24             22       NaN                   NaN       1      1987.0   18727   
25             23       NaN                   NaN       1      1969.0   14982   
26             24       NaN                   NaN       1      1957.0   16391   
27             25       NaN                   NaN       1      1981.0   15947   
28             26       NaN                   NaN       1      1993.0   16917   
29             27       NaN                   NaN       1      1975.0   15924   
...           ...       ...                   ...     ...         ...     ...   
46170       46171       NaN                   NaN       1      1988.0   15733   
46171       46172       NaN                   NaN       1      1995.0   16931   
46172       46173       NaN                   NaN       1      1987.0   17202   
46173       46174       NaN                   NaN       2      1989.0   20314   
46174       46175       NaN                   NaN       1      1989.0   21531   
46175       46176       NaN                   NaN       1      1968.0   21491   
46176       46177       NaN                   NaN       1      1990.0   15367   
46177       46178       NaN                   NaN       2      1973.0   18978   
46178       46179       NaN                   NaN       1      1980.0   19507   
46179       46180       NaN                   NaN       1      1980.0   16936   
46180       46181       NaN                   NaN       1      1982.0   18689   
46181       46182       NaN                   NaN       1      1972.0   15781   
46182       46183       NaN                   NaN       1      1988.0   15081   
46183       46184       NaN                   NaN       1      1983.0   21672   
46184       46185       NaN                   NaN       1      1970.0   16336   
46185       46186       NaN                   NaN       1      1973.0   21609   
46186       46187       NaN                   NaN       1      1964.0   19470   
46187       46188       NaN           

# Task 1 — Familiarize with SQL Clauses


•Sort data by start_station_id, tripduration

- Only checking trips with duration <= 3 hours

• Only show the top/last 10 records (aka head and tail in SQL)

• List all unique start_station_id values

• Aggregation functions:
- Count the number of trips (aka wc -l in SQL)
- Find the average/min/max trip duratio

In [6]:
#Sort data by start_station_id, tripduration
#- Only checking trips with duration <= 3 hours 

query_1 = '''
SELECT * FROM citibike
WHERE tripduration <= 10800
ORDER BY start_station_id, tripduration'''

pd.read_csv(StringIO.StringIO(queryCartoDB(query_1)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0            8998       NaN                   NaN       2      1980.0   19039   
1           40759       NaN                   NaN       2      1959.0   18131   
2           40757       NaN                   NaN       1      1995.0   19729   
3           21948       NaN                   NaN       1      1970.0   16879   
4           43099       NaN                   NaN       1      1983.0   15277   
5           32566       NaN                   NaN       1      1971.0   18492   
6           15971       NaN                   NaN       1      1961.0   19216   
7           41355       NaN                   NaN       1      1974.0   21588   
8           42074       NaN                   NaN       1      1978.0   19124   
9            2282       NaN                   NaN       1      1974.0   18174   
10          44674       NaN                   NaN       1      1983.0   18198   
11           2142       NaN                   NaN       1      1982.0   19672   
12            649       NaN                   NaN       2      1986.0   14692   
13          44789       NaN                   NaN       1      1983.0   17669   
14           9494       NaN                   NaN       1      1961.0   16754   
15          42446       NaN                   NaN       1      1983.0   21114   
16          41465       NaN                   NaN       1      1970.0   19729   
17          14395       NaN                   NaN       1      1981.0   18620   
18          19984       NaN                   NaN       1      1981.0   20119   
19           8726       NaN                   NaN       1      1981.0   20988   
20          25842       NaN                   NaN       2      1972.0   16310   
21          26155       NaN                   NaN       1      1978.0   18287   
22          25587       NaN                   NaN       1      1973.0   18991   
23          24156       NaN                   NaN       1      1981.0   20600   
24          27104       NaN                   NaN       1      1973.0   15927   
25           3957       NaN                   NaN       1      1991.0   15282   
26           9194       NaN                   NaN       1      1980.0   17196   
27          31639       NaN                   NaN       1      1978.0   16405   
28          38564       NaN                   NaN       1      1974.0   19771   
29          33221       NaN                   NaN       2      1986.0   17972   
...           ...       ...                   ...     ...         ...     ...   
46121       31410       NaN                   NaN       1      1958.0   17595   
46122       22533       NaN                   NaN       1      1962.0   15178   
46123        1333       NaN                   NaN       0         NaN   18288   
46124       23807       NaN                   NaN       1      1974.0   16147   
46125       27637       NaN                   NaN       1      1973.0   17061   
46126       22399       NaN                   NaN       1      1991.0   17198   
46127       24714       NaN                   NaN       1      1958.0   20872   
46128       41541       NaN                   NaN       1      1955.0   15495   
46129       43007       NaN                   NaN       1      1966.0   20780   
46130       35349       NaN                   NaN       1      1956.0   17077   
46131       20877       NaN                   NaN       1      1986.0   20268   
46132       39913       NaN                   NaN       1      1954.0   15586   
46133       38954       NaN                   NaN       1      1986.0   19331   
46134       30026       NaN                   NaN       1      1949.0   15627   
46135       14538       NaN                   NaN       1      1959.0   17488   
46136        2368       NaN                   NaN       0         NaN   16918   
46137       24145       NaN                   NaN       1      1959.0   15102   
46138       32820       NaN           

In [10]:
#Only show the top/last 10 records (aka head and tail in SQL) 

query_2 = '''
SELECT * FROM citibike
WHERE tripduration <= 10800
ORDER BY start_station_id ASC, tripduration ASC
LIMIT 10'''


query_3 = '''SELECT * FROM citibike
WHERE tripduration <= 10800
ORDER BY start_station_id DESC, tripduration DESC
LIMIT 10'''

pd.read_csv(StringIO.StringIO(queryCartoDB(query_2)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0        8998       NaN                   NaN       2        1980   19039   
1       40759       NaN                   NaN       2        1959   18131   
2       40757       NaN                   NaN       1        1995   19729   
3       21948       NaN                   NaN       1        1970   16879   
4       43099       NaN                   NaN       1        1983   15277   
5       32566       NaN                   NaN       1        1971   18492   
6       15971       NaN                   NaN       1        1961   19216   
7       41355       NaN                   NaN       1        1974   21588   
8       42074       NaN                   NaN       1        1978   19124   
9        2282       NaN                   NaN       1        1974   18174   

   end_station_id  start_station_id  tripduration    usertype  \
0              72                72           107  Subscriber   
1              72                72           189  Subscriber   
2              72                72           229  Subscriber   
3             478                72           248  Subscriber   
4             515                72           252  Subscriber   
5             478                72           261  Subscriber   
6             478                72           276  Subscriber   
7             447                72           283  Subscriber   
8             514                72           288  Subscriber   
9             447                72           301  Subscriber   

   end_station_name start_station_name  end_station_longitude  \
0  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
1  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
2  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
3  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
4  W 43 St & 10 Ave   W 52 St & 11 Ave             -73.994618   
5  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
6  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
7   8 Ave & W 52 St   W 52 St & 11 Ave             -73.985162   
8  12 Ave & W 40 St   W 52 St & 11 Ave             -74.002777   
9   8 Ave & W 52 St   W 52 St & 11 Ave             -73.985162   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.767272               -73.993929               40.767272   
1             40.767272               -73.993929               40.767272   
2             40.767272               -73.993929               40.767272   
3             40.760301               -73.993929               40.767272   
4             40.760094               -73.993929               40.767272   
5             40.760301               -73.993929               40.767272   
6             40.760301               -73.993929               40.767272   
7             40.763707               -73.993929               40.767272   
8             40.760875               -73.993929               40.767272   
9             40.763707               -73.993929               40.767272   

                 stoptime               starttime  
0  2015-02-03 10:12:00+00  2015-02-03 10:10:00+00  
1  2015-02-07 10:43:00+00  2015-02-07 10:40:00+00  
2  2015-02-07 10:44:00+00  2015-02-07 10:40:00+00  
3  2015-02-05 06:24:00+00  2015-02-05 06:19:00+00  
4  2015-02-07 15:47:00+00  2015-02-07 15:43:00+00  
5  2015-02-06 09:02:00+00  2015-02-06 08:58:00+00  
6  2015-02-04 12:51:00+00  2015-02-04 12:46:00+00  
7  2015-02-07 12:15:00+00  2015-02-07 12:10:00+00  
8  2015-02-07 13:54:00+00  2015-02-07 13:50:00+00  
9  2015-02-01 13:13:00+00  2015-02-01 13:08:00+00

In [11]:
pd.read_csv(StringIO.StringIO(queryCartoDB(query_3)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0        1506       NaN                   NaN       0         NaN   21202   
1        8831       NaN                   NaN       1      1958.0   15552   
2       23529       NaN                   NaN       1      1985.0   21473   
3       10173       NaN                   NaN       1      1969.0   15309   
4       27203       NaN                   NaN       1      1969.0   20069   
5        6781       NaN                   NaN       1      1959.0   19665   
6       44703       NaN                   NaN       2      1964.0   15443   
7       11587       NaN                   NaN       1      1996.0   19133   
8        1228       NaN                   NaN       0         NaN   17849   
9       17590       NaN                   NaN       1      1969.0   20246   

   end_station_id  start_station_id  tripduration    usertype  \
0             499              3002          7225    Customer   
1             510              3002          2139  Subscriber   
2             359              3002          2131  Subscriber   
3             318              3002          1977  Subscriber   
4             318              3002          1977  Subscriber   
5             494              3002          1976  Subscriber   
6             410              3002          1920  Subscriber   
7             279              3002          1908  Subscriber   
8             525              3002          1900    Customer   
9             318              3002          1830  Subscriber   

           end_station_name          start_station_name  \
0        Broadway & W 60 St  South End Ave & Liberty St   
1           W 51 St & 6 Ave  South End Ave & Liberty St   
2        E 47 St & Park Ave  South End Ave & Liberty St   
3  E 43 St & Vanderbilt Ave  South End Ave & Liberty St   
4  E 43 St & Vanderbilt Ave  South End Ave & Liberty St   
5           W 26 St & 8 Ave  South End Ave & Liberty St   
6   Suffolk St & Stanton St  South End Ave & Liberty St   
7      Peck Slip & Front St  South End Ave & Liberty St   
8          W 34 St & 11 Ave  South End Ave & Liberty St   
9  E 43 St & Vanderbilt Ave  South End Ave & Liberty St   

   end_station_longitude  end_station_latitude  start_station_longitude  \
0             -73.981918             40.769155               -74.015756   
1             -73.980420             40.760660               -74.015756   
2             -73.974987             40.755103               -74.015756   
3             -73.977987             40.753202               -74.015756   
4             -73.977987             40.753202               -74.015756   
5             -73.997236             40.747348               -74.015756   
6             -73.985180             40.720664               -74.015756   
7             -74.001670             40.707873               -74.015756   
8             -74.002116             40.755942               -74.015756   
9             -73.977987             40.753202               -74.015756   

   start_station_latitude                stoptime               starttime  
0               40.711512  2015-02-01 13:47:00+00  2015-02-01 11:46:00+00  
1               40.711512  2015-02-03 10:09:00+00  2015-02-03 09:33:00+00  
2               40.711512  2015-02-05 09:17:00+00  2015-02-05 08:42:00+00  
3               40.711512  2015-02-03 16:37:00+00  2015-02-03 16:04:00+00  
4               40.711512  2015-02-05 16:44:00+00  2015-02-05 16:11:00+00  
5               40.711512  2015-02-02 09:53:00+00  2015-02-02 09:21:00+00  
6               40.711512  2015-02-07 19:08:00+00  2015-02-07 18:36:00+00  
7               40.711512  2015-02-03 19:26:00+00  2015-02-03 18:54:00+00  
8               40.711512  2015-02-01 11:45:00+00  2015-02-01 11:14:00+00  
9               40.711512  2015-02-04 17:04:00+00  2015-02-04 16:33:00+00

In [14]:
#List unique start station id values

query_4 = '''
SELECT DISTINCT(start_station_id)
AS distinct_station_id
FROM citibike
LIMIT 10'''

pd.read_csv(StringIO.StringIO(queryCartoDB(query_4)), sep=',')

distinct_station_id  Unnamed: 1
0                  195         NaN
1                  251         NaN
2                  285         NaN
3                  120         NaN
4                  483         NaN
5                  264         NaN
6                  497         NaN
7                  305         NaN
8                  318         NaN
9                  312         NaN

In [18]:
# Count number of trips and find min max avg trip duration

query_5 = '''
SELECT DISTINCT start_station_id,
COUNT(start_station_id) AS trip_count,
AVG(tripduration) AS trip_avg,
MAX(tripduration) AS trip_max,
MIN(tripduration) AS trip_min
FROM citibike
WHERE tripduration <=10800
GROUP BY start_station_id
LIMIT 10'''

pd.read_csv(StringIO.StringIO(queryCartoDB(query_5)), sep=',')

start_station_id  trip_count    trip_avg  trip_max  trip_min
0               270          92  422.163043      2295       120
1               460          34  574.176471      2127        80
2               392          59  616.372881      1931       147
3               262          37  401.324324      1223       186
4               508          57  815.175439      1791       274
5               223         179  546.357542      2780        83
6               474         230  526.426087      3220        60
7               232          22  634.772727      1893        91
8               483         229  631.043668      8470       124
9               261          50  878.120000      6349       152

# Task 2 — Working with date/time

• Selecting trips started on Feb-02-2015 only

• Selecting trips started on the weekends
   
    - What are average trip duration during weekends?
• Can we do the same for weekday

In [26]:
#Feb 02 2015 only

query_6 = '''
SELECT * FROM citibike
WHERE starttime >= '2015-02-02 00:00'
AND starttime < '2015-02-03 00:00'
LIMIT 10'''

#trips that start on weekends

query_7 = '''
SELECT * FROM citibike
WHERE EXTRACT(DOW FROM starttime) IN (0,6)
LIMIT 10'''

# average trip duration during weekends

query_8 = '''
SELECT start_station_id,
starttime,
usertype,
gender,
AVG(tripduration) AS trip_avg
FROM citibike
WHERE EXTRACT(DOW FROM starttime) IN (0,6)
GROUP BY start_station_id, starttime, usertype, gender
LIMIT 10'''

# average trip duration during weekdays

query_9 = '''
SELECT start_station_id,
starttime,
usertype,
gender,
AVG(tripduration) AS trip_avg
FROM citibike
WHERE EXTRACT(DOW FROM starttime) NOT IN (0,6)
GROUP BY start_station_id, starttime, usertype, gender
LIMIT 10'''

pd.read_csv(StringIO.StringIO(queryCartoDB(query_6)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0        6442       NaN                   NaN       1        1992   20684   
1        6443       NaN                   NaN       2        1964   16094   
2        6444       NaN                   NaN       1        1974   17362   
3        6445       NaN                   NaN       2        1992   15475   
4        6446       NaN                   NaN       1        1968   16272   
5        6447       NaN                   NaN       1        1979   16370   
6        6448       NaN                   NaN       1        1963   19437   
7        6449       NaN                   NaN       1        1982   21211   
8        6450       NaN                   NaN       2        1978   21417   
9        6451       NaN                   NaN       1        1956   15722   

   end_station_id  start_station_id  tripduration    usertype  \
0             489               442           199  Subscriber   
1             349               326           418  Subscriber   
2             415               279           276  Subscriber   
3             237               496           420  Subscriber   
4             450               477           304  Subscriber   
5             391               467           622  Subscriber   
6             482               434           371  Subscriber   
7            2008               330           477  Subscriber   
8             423               521           743  Subscriber   
9             453               504           599  Subscriber   

            end_station_name    start_station_name  end_station_longitude  \
0           10 Ave & W 28 St       W 27 St & 7 Ave             -74.001768   
1    Rivington St & Ridge St       E 11 St & 1 Ave             -73.983299   
2  Pearl St & Hanover Square  Peck Slip & Front St             -74.009260   
3            E 11 St & 2 Ave       E 16 St & 5 Ave             -73.986724   
4            W 49 St & 8 Ave       W 41 St & 8 Ave             -73.987882   
5        Clark St & Henry St       Dean St & 4 Ave             -73.993446   
6            W 15 St & 7 Ave       9 Ave & W 18 St             -73.999318   
7      Little West St & 1 Pl   Reade St & Broadway             -74.016777   
8            W 54 St & 9 Ave       8 Ave & W 31 St             -73.986905   
9            W 22 St & 8 Ave       1 Ave & E 15 St             -73.999154   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.750664               -73.993915               40.746647   
1             40.718502               -73.984267               40.729538   
2             40.704718               -74.001670               40.707873   
3             40.730473               -73.992390               40.737262   
4             40.762272               -73.990026               40.756405   
5             40.697601               -73.978951               40.683125   
6             40.739355               -74.003664               40.743174   
7             40.705693               -74.005628               40.714505   
8             40.765849               -73.994811               40.750450   
9             40.744751               -73.981656               40.732219   

                 stoptime               starttime  
0  2015-02-02 00:05:00+00  2015-02-02 00:02:00+00  
1  2015-02-02 00:10:00+00  2015-02-02 00:03:00+00  
2  2015-02-02 00:09:00+00  2015-02-02 00:04:00+00  
3  2015-02-02 00:13:00+00  2015-02-02 00:06:00+00  
4  2015-02-02 00:12:00+00  2015-02-02 00:07:00+00  
5  2015-02-02 00:18:00+00  2015-02-02 00:08:00+00  
6  2015-02-02 00:15:00+00  2015-02-02 00:09:00+00  
7  2015-02-02 00:19:00+00  2015-02-02 00:11:00+00  
8  2015-02-02 00:25:00+00  2015-02-02 00:12:00+00  
9  2015-02-02 00:22:00+00  2015-02-02 00:12:00+00

In [23]:
pd.read_csv(StringIO.StringIO(queryCartoDB(query_7)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0        5960       NaN                   NaN       1        1965   19217   
1           1       NaN                   NaN       2        1978   17131   
2           2       NaN                   NaN       1        1993   21289   
3           3       NaN                   NaN       2        1969   18903   
4           4       NaN                   NaN       2        1985   21044   
5           5       NaN                   NaN       1        1957   19868   
6           6       NaN                   NaN       1        1979   15854   
7           7       NaN                   NaN       1        1983   15173   
8           8       NaN                   NaN       1        1955   17862   
9           9       NaN                   NaN       2        1985   21183   

   end_station_id  start_station_id  tripduration    usertype  \
0             462               435           351  Subscriber   
1             423               521           801  Subscriber   
2             504               497           379  Subscriber   
3             127               281          2474  Subscriber   
4             505              2004           818  Subscriber   
5              83               323           544  Subscriber   
6            2002               373           717  Subscriber   
7             504               352          1306  Subscriber   
8             116               439           913  Subscriber   
9            2012               335           759  Subscriber   

                end_station_name                 start_station_name  \
0               W 22 St & 10 Ave                    W 21 St & 6 Ave   
1                W 54 St & 9 Ave                    8 Ave & W 31 St   
2                1 Ave & E 15 St                 E 17 St & Broadway   
3          Barrow St & Hudson St  Grand Army Plaza & Central Park S   
4                6 Ave & W 33 St                  6 Ave & Broome St   
5  Atlantic Ave & Fort Greene Pl        Lawrence St & Willoughby St   
6   Wythe Ave & Metropolitan Ave       Willoughby Ave & Walworth St   
7                1 Ave & E 15 St                    W 56 St & 6 Ave   
8                W 17 St & 8 Ave                     E 4 St & 2 Ave   
9                E 27 St & 1 Ave           Washington Pl & Broadway   

   end_station_longitude  end_station_latitude  start_station_longitude  \
0             -74.004519             40.746920               -73.994156   
1             -73.986905             40.765849               -73.994811   
2             -73.981656             40.732219               -73.990093   
3             -74.006744             40.731724               -73.973715   
4             -73.988484             40.749013               -74.004704   
5             -73.976323             40.683826               -73.986317   
6             -73.963198             40.716887               -73.953820   
7             -73.981656             40.732219               -73.977225   
8             -74.001497             40.741776               -73.989780   
9             -73.976806             40.739445               -73.994046   

   start_station_latitude                stoptime               starttime  
0               40.741740  2015-02-01 20:42:00+00  2015-02-01 20:36:00+00  
1               40.750450  2015-02-01 00:14:00+00  2015-02-01 00:00:00+00  
2               40.737050  2015-02-01 00:07:00+00  2015-02-01 00:00:00+00  
3               40.764397  2015-02-01 00:42:00+00  2015-02-01 00:01:00+00  
4               40.724399  2015-02-01 00:15:00+00  2015-02-01 00:01:00+00  
5               40.692362  2015-02-01 00:10:00+00  2015-02-01 00:01:00+00  
6               40.693317  2015-02-01 00:14:00+00  2015-02-01 00:02:00+00  
7               40.763406  2015-02-01 00:26:00+00  2015-02-01 00:04:00+00  
8               40.726281  2015-02-01 00:19:00+00  2015-02-01 00:04:00+00  
9               40.729039  2015-02-01 00:17:00+00  2015-02-01 00:04:00+

In [25]:
pd.read_csv(StringIO.StringIO(queryCartoDB(query_8)), sep=',')

start_station_id               starttime    usertype  gender  trip_avg
0               336  2015-02-07 18:07:00+00  Subscriber       2     636.0
1               270  2015-02-01 11:35:00+00  Subscriber       1     387.0
2               161  2015-02-07 10:36:00+00  Subscriber       1     337.0
3               379  2015-02-01 13:52:00+00  Subscriber       1     408.0
4               271  2015-02-07 11:59:00+00  Subscriber       1     461.0
5               318  2015-02-07 06:45:00+00  Subscriber       1     426.0
6               488  2015-02-07 20:02:00+00  Subscriber       1    2053.0
7               262  2015-02-01 15:18:00+00  Subscriber       1     620.0
8               116  2015-02-07 19:28:00+00  Subscriber       1    1176.0
9               475  2015-02-07 21:48:00+00  Subscriber       1     391.0

In [27]:
pd.read_csv(StringIO.StringIO(queryCartoDB(query_9)), sep=',')

start_station_id               starttime    usertype  gender  trip_avg
0               415  2015-02-03 19:11:00+00  Subscriber       1     585.0
1               445  2015-02-03 12:14:00+00  Subscriber       2     299.0
2               383  2015-02-04 06:34:00+00  Subscriber       1     395.0
3               116  2015-02-04 19:07:00+00  Subscriber       1     648.0
4               498  2015-02-06 09:35:00+00  Subscriber       1     121.0
5               348  2015-02-06 13:47:00+00  Subscriber       1     469.0
6               229  2015-02-05 13:28:00+00  Subscriber       2     226.0
7               537  2015-02-02 16:05:00+00  Subscriber       1    1539.0
8               518  2015-02-05 20:26:00+00  Subscriber       1     632.0
9               487  2015-02-06 07:45:00+00  Subscriber       1     253.0

# Task 3 — Working with Space

• Showing the list of start station locations
- Using GROUP BY

• Showing the number of trips started per station

• … but only for stations within 500m of Time Square!

- The coordinates of Time Square is (40.7577,-73.9857)


In [33]:
# Station locations via groupby

query_10 ='''
SELECT CDB_TransformToWebmercator(CDB_LatLng(
start_station_latitude,
start_station_longitude)) 
AS THE_GEOM_WEBMERCATOR,
MIN(cartodb_id) as cartodb_id
FROM citibike

GROUP BY start_station_latitude, start_station_longitude
LIMIT 10
'''

# Number of trips started at each station

query_11 = '''
SELECT start_station_id,
CDB_TRANSFORMTOWEBMERCATOR(CDB_LatLng(
start_station_latitude,
start_station_longitude
))
AS THE_GEOM_WEBMERCATOR,
MIN(cartodb_id) AS carto_id,
COUNT(tripduration) AS trip_count
FROM citibike

GROUP BY start_station_id, start_station_latitude,
start_station_longitude
LIMIT 10'''


# Showing the number of trips started per station
# but only for stations within 500m of Time Square!
# The coordinates of Time Square is (40.7577,-73.9857)

query_12 = '''
SELECT start_station_id,
CDB_TRANSFORMTOWEBMERCATOR(CDB_LatLng(
start_station_latitude,
start_station_longitude
))
AS THE_GEOM_WEBMERCATOR,
MIN(cartodb_id) AS carto_id,
COUNT(tripduration) AS trip_count
FROM citibike
WHERE ST_DWITHIN(CDB_LatLng(
start_station_latitude,
start_station_longitude)::GEOGRAPHY,
CDB_LatLng(40.7577, -73.9857)::GEOGRAPHY, 500
)

GROUP BY start_station_id, start_station_latitude, start_station_longitude
LIMIT 10'''

pd.read_csv(StringIO.StringIO(queryCartoDB(query_10)), sep=',')

the_geom_webmercator  cartodb_id
0  0101000020110F0000863F88A4096A5FC19D1DAFD5DFF1...          84
1  0101000020110F0000BB2EC9F82D6A5FC19150E1B461FB...         412
2  0101000020110F0000D62721D3CA6D5FC157A5D529F6F5...         516
3  0101000020110F000060197671C96C5FC197A854AF46FA...         359
4  0101000020110F0000993A68067B6C5FC1B222DA492BFA...          78
5  0101000020110F00004FF9F3E90D6D5FC1F284479FC8F9...        1019
6  0101000020110F00003816CC930D6B5FC17B393D28A0F9...         110
7  0101000020110F00006B6C9CBF186A5FC109C57CA0D8FC...           7
8  0101000020110F00008B8D725EB16B5FC1BF41718A0DF6...         460
9  0101000020110F0000BA97F1D5D36C5FC13D326BC3E4F5...         799

In [32]:
pd.read_csv(StringIO.StringIO(queryCartoDB(query_11)), sep=',')

start_station_id                               the_geom_webmercator  \
0               438  0101000020110F0000B77CA834036B5FC19EF5005EBCF7...   
1               342  0101000020110F0000BEA7E6966A6A5FC1E90E1AC83EF6...   
2               242  0101000020110F0000E097312EB1695FC111C4474A72F3...   
3               161  0101000020110F00009B8267C45D6C5FC1577C6302EFF7...   
4               373  0101000020110F0000C12D0F658D675FC1323CC0ADCAF2...   
5               168  0101000020110F000071CD414CFB6B5FC149C2933772F9...   
6               502  0101000020110F00002877F4708B6A5FC1ED3B00D9C9F5...   
7               421  0101000020110F00000C641AC573695FC10D6A916323F3...   
8               511  0101000020110F000049C345A6266A5FC17E86BFF6F6F7...   
9               422  0101000020110F0000CECBC1AD456B5FC149916BC5DDFD...   

   carto_id  trip_count  
0        79         156  
1       609          78  
2       454          46  
3       185         203  
4         6          27  
5        49         327  
6       289         151  
7       238          37  
8        69         279  
9      1386         100

In [34]:
pd.read_csv(StringIO.StringIO(queryCartoDB(query_12)), sep=',')

start_station_id                               the_geom_webmercator  \
0               477  0101000020110F0000F55695027D6B5FC186E5BB69D7FB...   
1              2021  0101000020110F00001607D538556B5FC119A6CA6F41FC...   
2               465  0101000020110F000004AF971C1D6B5FC17A48F3C1A8FB...   
3               493  0101000020110F00009BB87D02B76A5FC146C144E9E5FB...   
4               524  0101000020110F0000F22A632FBE6A5FC1363A28CFADFB...   
5               529  0101000020110F00003F9300B2976B5FC1B8B88F3102FC...   
6               173  0101000020110F000020D6742CE16A5FC112F4D73D73FC...   

   carto_id  trip_count  
0        33         507  
1       124         141  
2        54         251  
3        19         141  
4      1115         112  
5       783         221  
6        42         213

# Task 4 — Putting it all together

• Find the station that had the longest average trip duration during
weekends and within 500m of TimeSquare!


• Extra: create lines for trips started from stations within 500m of Times
Squares and lasted less than 2 hours. The number of trips per each
pair of stations are output as attributes of these lines. 

In [37]:
query_13 = '''SELECT start_station_id,
starttime,
usertype,
gender,
AVG(tripduration) AS trip_avg
FROM citibike
WHERE EXTRACT(DOW FROM starttime) IN (0,6)
AND ST_DWITHIN(CDB_LatLng(
start_station_latitude,
start_station_longitude)::GEOGRAPHY,
CDB_LatLng(40.7577, -73.9857)::GEOGRAPHY, 500
)
GROUP BY start_station_id, starttime, usertype, gender
ORDER BY trip_avg DESC
LIMIT 10'''

In [38]:
pd.read_csv(StringIO.StringIO(queryCartoDB(query_13)), sep=',')

start_station_id               starttime    usertype  gender  trip_avg
0               173  2015-02-07 17:23:00+00  Subscriber       1   15636.0
1              2021  2015-02-07 18:28:00+00  Subscriber       1    3517.0
2               524  2015-02-07 14:28:00+00    Customer       0    2811.0
3               173  2015-02-07 16:21:00+00  Subscriber       1    2797.0
4              2021  2015-02-01 17:56:00+00  Subscriber       1    2203.0
5               493  2015-02-07 21:29:00+00  Subscriber       2    2200.0
6               477  2015-02-07 16:39:00+00  Subscriber       1    2192.0
7               529  2015-02-07 10:15:00+00  Subscriber       1    1959.0
8               477  2015-02-07 14:30:00+00  Subscriber       1    1957.0
9              2021  2015-02-07 00:16:00+00  Subscriber       2    1930.0

# Station number 173 has the longest average trip duration during weekends and within 500m of Times Square. The trip duration average is 4.3 hours.

# extra credit: make the function python 2 and 3 compatible so that it works on the  PUI2016_Python3 kernel